### PR TITLE
Change to use file to store offset information instead of table

### PR DESCRIPTION
**Pull request by chtyim**
- Big refactoring to consolidate highly duplicated code between
  Kafka 0.8 and Kafka 0.10 implementations
- Introduced a common module for the common implementation

### CODE REVIEW

File: kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/batch/source/Kafka10Reader.java [hunk 1]
```diff
@@ -50,61 +50,62 @@
...
+    if (!hasNext()) {
+      throw new NoSuchElementException("No message is available");
     }
+
+    ConsumerRecord<byte[], byte[]> consumerRecord = messageIter.next();
+
+    byte[] keyBytes = consumerRecord.key();
+    byte[] value = consumerRecord.value();
+    if (value == null) {
+      LOG.warn("Received message with null message.payload with topic {} and partition {}",
```
Comment (albertshau) [inline kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/batch/source/Kafka10Reader.java]: was in existing code, but this is better as:

Received message with null payload from partition {} of topic {}.

This also seems like an easy way to flood the log with thousands of warnings. Do you know if the value is ever expected to be null?

File: kafka-plugins-common/src/main/java/co/cask/hydrator/plugin/batch/source/KafkaPartitionOffsets.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+  /**
+   * Loads the {@link KafkaPartitionOffsets} from the given input file.
+   *
+   * @param fc the hadoop {@link FileContext} to read from the input file
+   * @param offsetFile the input file
+   * @return a {@link KafkaPartitionOffsets} object decoded from the file
+   * @throws IOException if failed to read the file
+   */
+  public static KafkaPartitionOffsets load(FileContext fc, Path offsetFile) throws IOException {
```
Comment (albertshau) [inline kafka-plugins-common/src/main/java/co/cask/hydrator/plugin/batch/source/KafkaPartitionOffsets.java:65]: annotate offsetFile as nullable

File: kafka-plugins-common/src/test/java/co/cask/hydrator/AbstractKafkaBatchSourceTest.java [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    sourceProperties.put("topic", "users");
+    sourceProperties.put("schema", schema.toString());
+    sourceProperties.put("format", "csv");
+    sourceProperties.put("maxNumberRecords", "${maxNumberRecords}");
+    ETLStage source = new ETLStage("source", new ETLPlugin(getKafkaBatchSourceName(),
+                                                           BatchSource.PLUGIN_TYPE, sourceProperties, null));
+    ETLStage sink = new ETLStage("sink", MockSink.getPlugin(outputName));
+
+    ETLBatchConfig pipelineConfig = ETLBatchConfig.builder()
+      .setTimeSchedule("* * * * *")
```
Comment (albertshau) [inline kafka-plugins-common/src/test/java/co/cask/hydrator/AbstractKafkaBatchSourceTest.java]: nit: this is not needed

File: kafka-plugins-common/src/test/java/co/cask/hydrator/AbstractKafkaBatchSourceTest.java [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    Map<String, String> messages = new HashMap<>();
+    messages.put("a", "1,samuel,jackson");
+    messages.put("b", "2,dwayne,johnson");
+    messages.put("c", "3,christopher,walken");
+    sendKafkaMessage("users", messages);
+
+
+    WorkflowManager workflowManager = appManager.getWorkflowManager(SmartWorkflow.NAME);
+    workflowManager.start(Collections.singletonMap("maxNumberRecords", "-1"));
+    workflowManager.waitForRun(ProgramRunStatus.COMPLETED, 1, TimeUnit.MINUTES);
```
Comment (albertshau) [inline kafka-plugins-common/src/test/java/co/cask/hydrator/AbstractKafkaBatchSourceTest.java]: can use startAndWaitForRun() to simplify a little bit and protect against changes in the future if somebody runs the program before this line.

File: kafka-plugins-common/src/main/java/co/cask/hydrator/plugin/batch/source/KafkaBatchConfig.java [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+  @Description("List of Kafka brokers specified in host1:port1,host2:port2 form. For example, " +
+    "host1.example.com:9092,host2.example.com:9092.")
+  @Macro
+  private String kafkaBrokers;
+
+  @Description("A directory path to store the latest Kafka offsets. " +
+    "A file named with the pipeline name will be created under the given directory.")
+  @Macro
+  @Nullable
+  private String offsetDir;
```
Comment (albertshau) [inline kafka-plugins-common/src/main/java/co/cask/hydrator/plugin/batch/source/KafkaBatchConfig.java:45]: widget files and reference docs need to be updated to include this new property.

**Review by albertshau — COMMENTED** [review]
minor comments, thanks for removing all the code duplication

File: kafka-plugins-common/src/main/java/co/cask/hydrator/plugin/batch/source/KafkaPartitionOffsets.java (hunk 2 above)
Comment (chtyim) [reply kafka-plugins-common/src/main/java/co/cask/hydrator/plugin/batch/source/KafkaPartitionOffsets.java:65]: Actually it never expected to be null and caller would make sure it is not null (if it is null, the caller shouldn't be able to construct the `FileContext`).

File: kafka-plugins-common/src/test/java/co/cask/hydrator/AbstractKafkaBatchSourceTest.java (hunk 4 above)
Comment (chtyim) [reply kafka-plugins-common/src/test/java/co/cask/hydrator/AbstractKafkaBatchSourceTest.java]: Oh, didn't know we have that.

File: kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/batch/source/Kafka10Reader.java (hunk 1 above)
Comment (chtyim) [reply kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/batch/source/Kafka10Reader.java]: Probably not logging at all.

Comment (chtyim) [reply kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/batch/source/Kafka10Reader.java]: Apparently it could be null based on the Kafka source.

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by chtyim** [comment]
Squashed and merging.

**Pull request merged**